src/test/java/com/uber/cadence/workflow/WorkflowTest.java

```diff
@@ -378,6 +378,55 @@ public void testActivityRetry() {
     assertEquals(activitiesImpl.toString(), 3, activitiesImpl.invocations.size());
   }
 
+  public static class TestActivityRetryOptionsChange implements TestWorkflow1 {
+
+    @Override
+    public String execute(String taskList) {
+      ActivityOptions.Builder options =
+          new ActivityOptions.Builder()
+              .setTaskList(taskList)
+              .setHeartbeatTimeout(Duration.ofSeconds(5))
+              .setScheduleToCloseTimeout(Duration.ofSeconds(5))
+              .setScheduleToStartTimeout(Duration.ofSeconds(5))
+              .setStartToCloseTimeout(Duration.ofSeconds(10));
+      if (Workflow.isReplaying()) {
+        options.setRetryOptions(
+            new RetryOptions.Builder()
+                .setMinimumAttempts(1)
+                .setMaximumInterval(Duration.ofSeconds(1))
+                .setInitialInterval(Duration.ofSeconds(1))
+                .setMaximumAttempts(3)
+                .build());
+      } else {
+        options.setRetryOptions(
+            new RetryOptions.Builder()
+                .setMinimumAttempts(2)
+                .setMaximumInterval(Duration.ofSeconds(1))
+                .setInitialInterval(Duration.ofSeconds(1))
+                .setMaximumAttempts(2)
+                .build());
+      }
+      TestActivities activities = Workflow.newActivityStub(TestActivities.class, options.build());
+      activities.throwIO();
+      return "ignored";
+    }
+  }
+
+  @Test
+  public void testActivityRetryOptionsChange() {
+    startWorkerFor(TestActivityRetryOptionsChange.class);
+    TestWorkflow1 workflowStub =
+        workflowClient.newWorkflowStub(
+            TestWorkflow1.class, newWorkflowOptionsBuilder(taskList).build());
+    try {
+      workflowStub.execute(taskList);
+      fail("unreachable");
+    } catch (WorkflowException e) {
+      assertTrue(e.getCause().getCause() instanceof IOException);
+    }
+    assertEquals(activitiesImpl.toString(), 3, activitiesImpl.invocations.size());
```

Comment:
This assertion seems wrong. MaxAttempt is set to 2 in the non-replay path and should be honored here.
Same below.